### ISSUE TITLE
[RFC] Who needs a non-python interface to starfish? (CLI?) 

### ISSUE THREAD

**Issue by ambrosejcarr** [spacetx/starfish#1288]
Research which users need/want the CLI.

**Comment by ttung** [comment]
I'm very curious about this.  I don't think I saw a single person use the CLI at the hackathon. :)

**Comment by dganguli** [comment]
Apologies for the naiive question! Even if users don't use the CLI to determine  _pipeline recipe parameters_ isn't the CLI still useful for taking a known pipeline recipe, and pipeline-ing them up? e.g., writing wdls, using snakemake, etc.

**Comment by ttung** [comment]
Strictly speaking, one can use the `starfish recipe` command to run a pipeline, and we can remove all the other CLI commands.

**Comment by ambrosejcarr** [comment]
>  isn't the CLI still useful for taking a known pipeline recipe, and pipeline-ing them up? e.g., writing wdls, using snakemake, etc

The CLI as written causes quite a lot of IO since it writes to disk after each command. Tony's recipe runner (mentioned above) is faster, and WDL can just execute python code directly, e.g. https://github.com/spacetx/starfish/blob/ajc-wdl-smfish/workflows/iss/iss.wdl

**Comment by ttung** [comment]
One of the key advantages of forcing recipe/CLI usage is that it enforces that everything is a pipeline component / algorithm, and eliminates a bunch of ad-hoc untraceable (i.e., no provenance data logged) python code.

**Comment by dganguli** [comment]
oh. thank you! @neuromusic originally the CLI was mean to support users who 1. were not comfortable with Python or 2. needed a CLI version to run a pipeline at scale given their compute infrastructure. 

In retrospect, 1 probably doesn't make sense? But it'd be good to get quantitative about that. I'm not sure how much 2. makes sense given @ambrosejcarr and @ttung s comments. But, I would ask consortia what they think?

**Comment by ambrosejcarr** [comment]
> But, I would ask consortia what they think?

👍 . Also anyone else we think might be key users of starfish.

**Comment by ambrosejcarr** [comment]
Here's a little manifesto about deleting the CLI. Please poke holes in it. I think if the team generally agrees with this then we should prioritize removing it. 

## Background

We've struggled to get a UX study off the ground to address this question. However, the decision to support the CLI adds substantial complexity to both internal development and external contribution. It forces us to generate boilerplate around very simple API calls such as reduction operations, produces technical burden for us to refactor, forces contributors to add multiple additional methods to add new algorithms, and increases the testing surface of starfish. 

By contrast, it is posed to add value for a group of users that are not confident users of a python API. Unfortunately, we haven't seen the CLI generate traction with users; in some cases those that aren't comfortable with python are also not comfortable with the command line. Finally, consortia and data generators can just as easily write a simple python script and run it in docker or singularity as they can a bash script based on a CLI. 

Because we have no current dependencies on the CLI, we have an opportunity to remove it, which will:
1. increase our velocity to work on new features 
2. decrease the time required to get feedback on feedback
3. expedite construction of an MVP product delivering images-to-expression matrix functionality
4. Reduce code size and complexity of new algorithm contributions

If it turns out we need a CLI in the future, It should be possible to write a separate click module that wraps all the functions without baking the auto-generation into the classes themselves. 

## Proposal

This part is a bit squishier and would benefit from more careful treatment from an engineer, but in broad strokes, a potential path forward could look like: 

* Remove the CLI
* Remove `PipelineComponent`
* Shift any needed boilerplate (e.g. logging) from pipeline component to `AlgorithmBase`
* Define an algorithm API that is as minimal as possible and keep it maximally consistent across AlgorithmBase subclasses. See [sklearn example](https://scikit-learn.org/stable/developers/contributing.html#apis-of-scikit-learn-objects). There are more extreme functional programming options that move the meta-programming into decorators, but the `sklearn` approach feels like a good intermediate non-nuclear option. ;-)

cc @shanaxel42 @ttung @neuromusic @dganguli @kne42 

Also @kevinyamauchi @berl @njmei If any of you secretly use the CLI, now would be the time to speak up. ;-)

**Comment by shanaxel42** [comment]
I'm pretty supportive of this. I think it would definitely decrease a lot of code complexity and barrier of entry for contributors. A few things:

 `Shift any needed boilerplate (e.g. logging) from pipeline component to AlgorithmBase`  the logging is already in AlgorithmBase

We would also have to remove all the nice runner code @ttung wrote, although I think your point about data generators being able to write python scripts instead is correct. 

This is a significant refactor. If we wanna go for it I think we should do it now before we start adding new 2.0 features.

**Comment by ambrosejcarr** [comment]
> Shift any needed boilerplate (e.g. logging) from pipeline component to AlgorithmBase the logging is already in AlgorithmBase

Oops. Thanks for the clarification. 

> We would also have to remove all the nice runner code @ttung wrote, although I think your point about data generators being able to write python scripts instead is correct.

😢 

> This is a significant refactor. If we wanna go for it I think we should do it now before we start adding new 2.0 features.

👍

**Comment by berl** [comment]
does this have any impact on scalability? I always thought one argument for the  CLI was that it provided some value for integration with existing pipeline/workflow tools. every time I talk about pipelines it eventually comes around to bash scripts, make, etc. and I think that's where the CLI would be helpful.  

Just to be clear: we're definitely in the API camp and we (ok, probably @njmei ) will make whatever scripts we need to automate our workflow.   My biggest concern in this part of the codebase is getting a runner together that can parallelize flexibly, store recipes in a sane manner and be kicked off with a one-liner.  If that one-liner is in a notebook or a script that's fine.  If dropping the CLI makes that happen faster, then I'm all for it.  I see that functionality as in-scope for `starfish` but maybe yall see that as the user's responsibility or for future work.

**Comment by ambrosejcarr** [comment]
> does this have any impact on scalability?

If you mean performance, API  = runner >> CLI I think. There should be no degradation. 

Otherwise you nailed all the reasons we want a CLI or runner in the longer term. The issue is that it introduces a lot of dev work while we're looking for high velocity in feature addition, and it presents a barrier to new contributions.

> If dropping the CLI makes that happen faster [...]

This is my hypothesis. *Adding the cli back later in a way that presents less work for comp bio developers is in scope*, I would just prefer to address that feature later when we have more users and can get more feedback on how it should work: right now we don't have any users who can leverage it because the package isn't adequately feature-complete -- most users need to implement _something_ to use the package. 

Since our users are devs _by requirement_, we may as well make the package work optimally for them!

**Comment by ttung** [comment]
One key advantage of the CLI / recipe interface is that it enforces us to structure everything as `AlgorithmBase` implementations.  Once we remove this restriction, people can just write a bunch of spaghetti python code that slice and dice the data however they want.  If that is the direction we go in, starfish mostly becomes a set of input and output file formats.  It's likely okay for starfish to exist like this, and there are still some distinct advantages over a world where everything, including the file formats, are nonstandard.  However, it's less clear to me how people could share algorithms.

**Comment by shanaxel42** [comment]
Not sure we really reached a consensus on this yesterday. @kevinyamauchi and I chatted a bit after the meeting as well so I wanted to propose a path forward. In general I think we all agree that the goal of this idea was to reduce barriers for contributing algorithms to starfish. But the discussion also raised valid concerns about it relaxing some of the api restrictions that enforce structure instead of unreproducible "spaghetti code". 

**Proposal**
Remove the CLI but maintain the pipeline component structure. 

- Remove the CLI
- Create good documentation on hot to add a new instance of a pipeline component class. 
- Potentially create a way to autogenerate some of the boilerplate component stuff

I think this would allow us to really focus on defining what the essential pieces of a pipeline are and communicate that to potential contributors. It also simplifies the code base considerably. We'd be able to remove all cli related code. The one downside this doesn't address is the comment @ttung made about potential memory leaks if our recipes our purely python. But my overall hypothesis is that really defining and making simple one way of creating a pipeline recipe instead of supporting multiple would do more to encourage people actually using the structure then having multiple options. This is just my hypothesis though, open to all the feedback @ambrosejcarr @ttung @berl @neuromusi @njmei

**Comment by ttung** [comment]
For what it's worth, recipes do manage memory slightly better than python scripts.  If it knows that a block of data is not going to be used any more during the recipe's execution, it will discard that data.  We could build even better memory management strategies if we get rid of in-place, though that could also inflate memory usage.

**Comment by shanaxel42** [comment]
#1398
